src/SignalR/clients/ts/signalr/src/MessageBuffer.ts

```diff
@@ -0,0 +1,218 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+import { IConnection } from "./IConnection";
+import { AckMessage, HubMessage, IHubProtocol, MessageType, SequenceMessage } from "./IHubProtocol";
+import { isArrayBuffer } from "./Utils";
+
+/** @private */
+export class MessageBuffer {
+    private readonly _protocol: IHubProtocol;
+    private readonly _connection: IConnection;
+
+    private readonly _bufferSize: number = 100_000;
+
+    private _messages: Key[] = [];
+    private _totalMessageCount: number = 0;
+    private _waitForSequenceMessage: boolean = false;
+
+    // Message IDs start at 1 and always increment by 1
+    private _currentReceivingSequenceId = 1;
+    private _latestReceivedSequenceId = -100;
+    private _bufferedByteCount: number = 0;
+
+    private _ackTimerHandle?: any;
+
+    private _backPressurePromise!: Promise<void>;
+    private _backpressurePromiseResolver: (value: void) => void = () => {};
+
+    private _resendPromise: Promise<void> = Promise.resolve();
+    private _resendResolve?: (value: void) => void = undefined;
+    private _resendReject?: (value?: any) => void = undefined;
+
+    constructor(protocol: IHubProtocol, connection: IConnection, bufferSize: number) {
+        this._protocol = protocol;
+        this._connection = connection;
+        this._bufferSize = bufferSize;
+    }
+
+    public async _send(message: HubMessage): Promise<void> {
+        if (this._bufferedByteCount >= this._bufferSize) {
+            await this._backPressurePromise;
```

Comment:
What if sends are not awaited? Could we continue to many awaiters at ounce without rechecking the `_bufferedByteCount`? I think we might need logic to allow a queue of awaiters to be released one at a time rather than all at once like [Kestrel's old HTTP/2 output flow control logic](https://github.com/dotnet/aspnetcore/blob/42af1fe8be1869b034feeec5aaefc0efb561564f/src/Servers/Kestrel/Core/src/Internal/Http2/FlowControl/OutputFlowControl.cs#L55-L58) used to handle streams when connection flow control was relieved. 

It might be easier to just not await before the `send`. I was about to argue that the message is already on the heap whether you await here or not, but that might hurt applications that would slow down the send rate so long as the receiver doesn't receive messages quickly due to some sort of application-level feedback. Something like a ping pong scenario would be like this. In such a scenario, delaying sends would effectively produce backpressure despite no promises being observed.